master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+	if err != nil {
+		return nil, err
+	}
+	if !s.AsyncHalvingConfig.SmallerIsBetter {
+		metric *= -1
+	}
+
+	return s.promoteAsync(ctx, requestID, message, metric), nil
+}
+
+func (s *asyncHalvingSearch) promoteAsync(
+	ctx context, requestID RequestID, message Workload, metric float64,
+) []Operation {
+	// Upon a trial is finished, we should return at least one more trial to train unless
+	// the bracket of successive halving is finished.
+	rungIndex := s.trialRungs[requestID]
+	rung := s.rungs[rungIndex]
+	rung.outstandingTrials--
+	addedNewTrial := false
+
+	ops := make([]Operation, 0)
+	// If the trial has completed the top rung's validation, close the trial.
+	if rungIndex == s.AsyncHalvingConfig.NumRungs-1 {
+		s.trialsCompleted++
+		if !s.earlyExitTrials[requestID] {
+			ops = append(ops, NewClose(requestID))
+		}
+	} else {
+		// This is not the top rung, so do promotions to the next rung.
+		nextRung := s.rungs[rungIndex+1]
+		for _, promotionID := range rung.promotionsAsync(
+			requestID,
+			metric,
+			s.AsyncHalvingConfig.Divisor,
+		) {
+			s.trialRungs[promotionID] = rungIndex + 1
+			nextRung.outstandingTrials++
+			if !s.earlyExitTrials[promotionID] {
+				ops = append(
+					ops, trainAndValidate(promotionID, rung.stepsNeeded, nextRung.stepsNeeded)...)
+				addedNewTrial = true
+			} else {
+				step := s.rungs[rungIndex+1].stepsNeeded
+				wkld := Workload{
+					Kind:         ComputeValidationMetrics,
+					ExperimentID: message.ExperimentID,
+					TrialID:      message.TrialID,
+					StepID:       step,
+				}
+
+				// We make a recursive call that will behave the same
+				// as if we'd actually run the promoted job and received
+				// the worse possible result in return.
+				return s.promoteAsync(ctx, promotionID, wkld, ashaExitedMetricValue)
+			}
+		}
+	}
+
+	allTrials := len(s.trialRungs)
+	if !addedNewTrial && allTrials < s.maxTrials {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		ops = append(ops, create)
+		ops = append(ops, trainAndValidate(create.RequestID, 0, s.rungs[0].stepsNeeded)...)
+		s.rungs[0].outstandingTrials++
+		s.trialRungs[create.RequestID] = 0
+	}
+
+	// Only close out trials once we have reached the maxTrials for the searcher.
+	if allTrials == s.maxTrials {
+		ops = append(ops, s.closeOutRungs()...)
+	}
+	return ops
+}
+
+// closeOutRungs closes all remaining unpromoted trials in any rungs that have no more outstanding
+// trials.
+func (s *asyncHalvingSearch) closeOutRungs() []Operation {
+	ops := make([]Operation, 0)
+	for _, rung := range s.rungs {
+		if rung.outstandingTrials > 0 {
+			break
+		}
+		for tid, trialMetric := range rung.metrics {
+			if !trialMetric.promoted && !trialMetric.closed {
+				s.trialsCompleted++
+				if !s.earlyExitTrials[trialMetric.requestID] {
+					ops = append(ops, NewClose(trialMetric.requestID))
+				}
+				rung.metrics[tid].closed = true
+				ops = append(ops, NewClose(trialMetric.requestID))
```

Comment:
blocking: I assume this is a typo? We wouldn't' want to send send a trial multiple close messages

Reply:
We won't be sending close op multiple times for a trial because of this check:  if !trialMetric.promoted && !trialMetric.closed
If a trial has metrics on multiple rungs, there will be only one rung in which it hasn't been promoted and the trial metric is not closed.

Reply:
Sorry, I could've been a bit clearer in my first message. L217 and L220 both do `ops = append(ops, NewClose(trialMetric.requestID))`. I definitely don't think we should be doing that. I think we should delete L220

Reply:
thanks for catching that!